### PR TITLE
Allow grouping of initiative for creatures of the same type

### PR DESCRIPTION
**Pull request by chrisbloom7**
#116 

Allow automatic grouping of initiative

- Adds an option under Rules to "Automatically group initiative rolls
- Auto grouping options are "None", "By Name", and "Side Initiative"
- "None" makes no changes to initiative as it works now
- "By Name" causes new non-player combatants to be linked automatically to other combatants of the same name
- "Side Initiative" causes players to be grouped with other players, and non-player combatants to be linked to non-player combatants. In this case, neither side recieves roll modifiers
- Auto groupings are not immutable. Users are still able to move combatants, unlink them, and relink them manually

### CODE REVIEW

File: client/Combatant/Combatant.ts [hunk 1]
```diff
@@ -168,7 +168,12 @@ export class Combatant implements Combatant {
         return modifiers;
     }
 
-    public GetInitiativeRoll = () => this.Encounter.Rules.AbilityCheck(this.InitiativeBonus, this.StatBlock().InitiativeAdvantage ? "advantage" : null);
+    public GetInitiativeRoll(): number {
```
Comment (cynicaloptimist) [inline client/Combatant/Combatant.ts]: Use arrow function syntax:
`public GetInitiativeRoll: () => number = () => {`
The benefit of that syntax is that it prevents the `this` object from being re-bound if you use this method as a free function.

File: client/Combatant/Combatant.ts [hunk 2]
```diff
@@ -23,7 +23,6 @@ export interface Combatant {
     InitiativeGroup: KnockoutObservable<string>;
     Hidden: KnockoutObservable<boolean>;
     StatBlock: KnockoutObservable<StatBlock>;
-    GetInitiativeRoll: () => number;
```
Comment (cynicaloptimist) [inline client/Combatant/Combatant.ts]: Why was this removed?

File: client/Settings/Settings.ts [hunk 3]
```diff
@@ -11,6 +11,7 @@ export interface Settings {
         RollMonsterHp: boolean;
         AllowNegativeHP: boolean;
         AutoCheckConcentration: boolean;
+        AutoGroupInitiative: string;
```
Comment (cynicaloptimist) [inline client/Settings/Settings.ts]: You can use a string literal type here to enforce that it's always a valid string:
`AutoGroupInitiative: "None" | "By Name" | "Side Initiative";`

There's a larger refactoring you could do here by encapsulating the AutoGroupInitiative options and behaviors into a class. Let me know if you're interested in taking that on, it would make the code a bit cleaner IMO and I can provide some guidance there.

File: client/Combatant/Combatant.ts (hunk 2 above)
Comment (chrisbloom7) [reply client/Combatant/Combatant.ts]: JSLint complained that it was defined twice after I made the changes to the function definition on line 171. I'll see if that changes after making the changes you requested. If not, do you want me to ignore the JSLint warning and leave it there?

File: client/Settings/Settings.ts (hunk 3 above)
Comment (chrisbloom7) [reply client/Settings/Settings.ts]: I had considered making the InitiativeGroup a class of its own, but that didn't seem to be worth while once I got a handle on what was needed. I could see making AutoGroupInitiative its own class thought. Maybe I can do that as a fast-follow PR? Any suggestions on where that class should live, i.e. folder of its own or alongside an existing class? I'd be open to hearing any other thoughts you had around it.

File: client/Combatant/Combatant.ts (hunk 1 above)
Comment (chrisbloom7) [reply client/Combatant/Combatant.ts]: I've immediately run into an error saying `Type 'string' is not assignable to type '"None" | "By Name" | "Side Initiative"'.` I see a number of open issues in the TypeScript repo about this. I think I could get around this by defining the list everywhere one of the values could be assigned, but not sure about that. Wondering instead if this would be better implemented as an Enum, and then referencing the enum values elsewhere in the code, possibly as part of the AutoGroupInitiative encapsulation mentioned below. Any ideas on this?

File: client/Settings/Settings.ts (hunk 3 above)
Comment (chrisbloom7) [reply client/Settings/Settings.ts]: I've immediately run into an error saying `Type 'string' is not assignable to type '"None" | "By Name" | "Side Initiative"'.` I see a number of open issues in the TypeScript repo about this. I think I could get around this by defining the list everywhere one of the values could be assigned, but not sure about that. Wondering instead if this would be better implemented as an Enum, and then referencing the enum values elsewhere in the code, possibly as part of the AutoGroupInitiative encapsulation. Any ideas on this?

Comment (cynicaloptimist) [reply client/Settings/Settings.ts]: Yeah, this isn't a TS bug, but it can be a bit hard to troubleshoot. Try extracting a type:
`export type AutoGroupInitiativeOption = "None" | "By Name" | "Side Initiative";`
Then use that type instead of string anywhere you're getting or setting this value:
`AutoGroupInitiative: AutoGroupInitiativeOption;`

I think hold off on adding a new class for `AutoGroupInitiative`. I'll think about a more general purpose solution to these types of user options, and your implementation is great as-is.

**Comment by cynicaloptimist** [comment]
Excellent, thanks Chris. I think it would make sense to start encapsulating some of our options concepts into classes. Other candidates for this treatment include `HpVerbosity` and possibly `AutoCheckConcentration`, or other options that require changes in program flow (rather than just display options).
